File: apps/entrelares_app/lib/models/family.dart
import 'package:entrelares_core/entrelares_core.dart';

/// The slice of the `families` row the entitlement mirror reads (F-32).
/// Mirrors `Entrelares/Models/Family.cs`: `plan` is written only by the
/// service-role RPC `set_family_plan`; `comp_premium_at` (F-58) only by
/// `admin_set_comp` — the client never writes any of this.
class Family {
  final int id;
  final String plan;
  final DateTime? trialEndsAt;
  final DateTime? compPremiumAt;

  const Family({
    required this.id,
    required this.plan,
    this.trialEndsAt,
    this.compPremiumAt,
  });

  factory Family.fromJson(Map<String, dynamic> json) => Family(
        id: json['id'] as int,
        plan: (json['plan'] as String?) ?? 'free',
        trialEndsAt: _utc(json['trial_ends_at'] as String?),
        compPremiumAt: _utc(json['comp_premium_at'] as String?),
      );

  static DateTime? _utc(String? wire) =>
      wire == null ? null : DateTime.parse(wire).toUtc();

  /// Mirror of `EntitlementService.IsPremium(Family?)`: null → free — the
  /// fail-closed default (an RLS-blocked or missing row never grants).
  static bool isPremiumFamily(Family? family, DateTime nowUtc) =>
      family != null &&
      computeIsPremium(
        plan: family.plan,
        trialEndsAtUtc: family.trialEndsAt,
        nowUtc: nowUtc,
        compPremiumAtUtc: family.compPremiumAt,
      );
}

File: db-gate/Entrelares.DbContracts/Models/Family.cs
using Supabase.Postgrest.Attributes;
using Supabase.Postgrest.Models;

namespace Entrelares.Models
{
    [Table("families")]
    public class Family : BaseModel
    {
        [PrimaryKey("id", false)]
        public long Id { get; set; }

        [Column("name")]
        public string Name { get; set; } = string.Empty;

        [Column("created_at")]
        public DateTime CreatedAt { get; set; }

        // F-32: freemium plan. "free" (default) or "premium". Written only by the
        // service-role RPC set_family_plan (never by the client).
        [Column("plan")]
        public string Plan { get; set; } = "free";

        // F-32: end of the 30-day Premium trial (UTC). Null on grandfathered
        // premium families and after a family drops back to free.
        [Column("trial_ends_at")]
        public DateTime? TrialEndsAt { get; set; }

        // F-58: permanent courtesy Premium (comp), granted only by the platform
        // operator's RPC (admin_set_comp). Orthogonal to plan/trial — a billing
        // downgrade never clears it. Null = no comp.
        [Column("comp_premium_at")]
        public DateTime? CompPremiumAt { get; set; }
    }
}
